### PR TITLE
#997 make incremental change capture columns more robust

### LINKED ISSUE #997: Running datacube system init on an existing database fails when adding new columns
### Expected behaviour
Updating the database structure should not cause the init process to fail

### Actual behaviour

An active database may have locking happening which causes the `datacube system init` command to fail. Logs:

> datacube -v system init
> 2020-07-24 01:28:43,559 362 datacube INFO Running datacube command: /env/bin/datacube -v system init
> Initialising database...
> 2020-07-24 01:28:43,808 362 datacube.drivers.postgres._core INFO Ensuring user roles.
> 2020-07-24 01:28:43,818 362 datacube.drivers.postgres._core INFO Adding role grants.
> 2020-07-24 01:28:43,839 362 datacube.drivers.postgres._core INFO Adding Update Triggers
> Traceback (most recent call last):
>   File "/env/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1278, in _execute_context
>     cursor, statement, parameters, context
>   File "/env/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 593, in do_execute
>     cursor.execute(statement, parameters)
> psycopg2.errors.DeadlockDetected: deadlock detected
> DETAIL:  Process 16680 waits for AccessExclusiveLock on relation 16492 of database 16592; blocked by process 18866.
> Process 18866 waits for AccessShareLock on relation 16431 of database 16592; blocked by process 16680.
> HINT:  See server log for query details.
> 
> 
> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last):
>   File "/env/bin/datacube", line 8, in <module>
>     sys.exit(cli())
>   File "/env/lib/python3.6/site-packages/click/core.py", line 829, in __call__
>     return self.main(*args, **kwargs)
>   File "/env/lib/python3.6/site-packages/click/core.py", line 782, in main
>     rv = self.invoke(ctx)
>   File "/env/lib/python3.6/site-packages/click/core.py", line 1259, in invoke
>     return _process_result(sub_ctx.command.invoke(sub_ctx))
>   File "/env/lib/python3.6/site-packages/click/core.py", line 1259, in invoke
>     return _process_result(sub_ctx.command.invoke(sub_ctx))
>   File "/env/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
>     return ctx.invoke(self.callback, **ctx.params)
>   File "/env/lib/python3.6/site-packages/click/core.py", line 610, in invoke
>     return callback(*args, **kwargs)
>   File "/env/lib/python3.6/site-packages/datacube/ui/click.py", line 196, in new_func
>     return f(parsed_config, *args, **kwargs)
>   File "/env/lib/python3.6/site-packages/datacube/ui/click.py", line 228, in with_index
>     return f(index, *args, **kwargs)
>   File "/env/lib/python3.6/site-packages/datacube/scripts/system.py", line 48, in database_init
>     with_permissions=init_users)
>   File "/env/lib/python3.6/site-packages/datacube/index/index.py", line 60, in init_db
>     is_new = self._db.init(with_permissions=with_permissions)
>   File "/env/lib/python3.6/site-packages/datacube/drivers/postgres/_connections.py", line 182, in init
>     _core.update_schema(self._engine)
>   File "/env/lib/python3.6/site-packages/datacube/drivers/postgres/_core.py", line 166, in update_schema
>     install_timestamp_trigger(c)
>   File "/env/lib/python3.6/site-packages/datacube/drivers/postgres/_triggers.py", line 50, in install_timestamp_trigger
>     conn.execute(UPDATE_COLUMN_MIGRATE_SQL_TEMPLATE.format(schema=SCHEMA_NAME, table=name))
>   File "/env/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1006, in execute
>     return self._execute_text(object_, multiparams, params)
>   File "/env/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1181, in _execute_text
>     parameters,
>   File "/env/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1318, in _execute_context
>     e, statement, parameters, cursor, context
>   File "/env/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1512, in _handle_dbapi_exception
>     sqlalchemy_exception, with_traceback=exc_info[2], from_=e
>   File "/env/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 178, in raise_
>     raise exception
>   File "/env/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1278, in _execute_context
>     cursor, statement, parameters, context
>   File "/env/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 593, in do_execute
>     cursor.execute(statement, parameters)
> sqlalchemy.exc.OperationalError: (psycopg2.errors.DeadlockDetected) deadlock detected
> DETAIL:  Process 16680 waits for AccessExclusiveLock on relation 16492 of database 16592; blocked by process 18866.
> Process 18866 waits for AccessShareLock on relation 16431 of database 16592; blocked by process 16680.
> HINT:  See server log for query details.
> 
> [SQL: 
> alter table agdc.dataset_source add column if not exists updated
> timestamptz not null default now();
> ]
> (Background on this error at: http://sqlalche.me/e/13/e3q8)
> jovyan@jupyter-alexgleith:~$ datacube -v system init
> 2020-07-24 01:45:37,650 386 datacube INFO Running datacube command: /env/bin/datacube -v system init
> Initialising database...
> 2020-07-24 01:45:37,901 386 datacube.drivers.postgres._core INFO Ensuring user roles.
> 2020-07-24 01:45:37,911 386 datacube.drivers.postgres._core INFO Adding role grants.
> 2020-07-24 01:45:37,918 386 datacube.drivers.postgres._core INFO Adding Update Triggers
> Traceback (most recent call last):
>   File "/env/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1278, in _execute_context
>     cursor, statement, parameters, context
>   File "/env/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 593, in do_execute
>     cursor.execute(statement, parameters)
> psycopg2.errors.DeadlockDetected: deadlock detected
> DETAIL:  Process 16605 waits for AccessExclusiveLock on relation 16492 of database 16592; blocked by process 18557.
> Process 18557 waits for AccessShareLock on relation 16431 of database 16592; blocked by process 16605.
> HINT:  See server log for query details.
> 
> 
> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last):
>   File "/env/bin/datacube", line 8, in <module>
>     sys.exit(cli())
>   File "/env/lib/python3.6/site-packages/click/core.py", line 829, in __call__
>     return self.main(*args, **kwargs)
>   File "/env/lib/python3.6/site-packages/click/core.py", line 782, in main
>     rv = self.invoke(ctx)
>   File "/env/lib/python3.6/site-packages/click/core.py", line 1259, in invoke
>     return _process_result(sub_ctx.command.invoke(sub_ctx))
>   File "/env/lib/python3.6/site-packages/click/core.py", line 1259, in invoke
>     return _process_result(sub_ctx.command.invoke(sub_ctx))
>   File "/env/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
>     return ctx.invoke(self.callback, **ctx.params)
>   File "/env/lib/python3.6/site-packages/click/core.py", line 610, in invoke
>     return callback(*args, **kwargs)
>   File "/env/lib/python3.6/site-packages/datacube/ui/click.py", line 196, in new_func
>     return f(parsed_config, *args, **kwargs)
>   File "/env/lib/python3.6/site-packages/datacube/ui/click.py", line 228, in with_index
>     return f(index, *args, **kwargs)
>   File "/env/lib/python3.6/site-packages/datacube/scripts/system.py", line 48, in database_init
>     with_permissions=init_users)
>   File "/env/lib/python3.6/site-packages/datacube/index/index.py", line 60, in init_db
>     is_new = self._db.init(with_permissions=with_permissions)
>   File "/env/lib/python3.6/site-packages/datacube/drivers/postgres/_connections.py", line 182, in init
>     _core.update_schema(self._engine)
>   File "/env/lib/python3.6/site-packages/datacube/drivers/postgres/_core.py", line 166, in update_schema
>     install_timestamp_trigger(c)
>   File "/env/lib/python3.6/site-packages/datacube/drivers/postgres/_triggers.py", line 50, in install_timestamp_trigger
>     conn.execute(UPDATE_COLUMN_MIGRATE_SQL_TEMPLATE.format(schema=SCHEMA_NAME, table=name))
>   File "/env/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1006, in execute
>     return self._execute_text(object_, multiparams, params)
>   File "/env/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1181, in _execute_text
>     parameters,
>   File "/env/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1318, in _execute_context
>     e, statement, parameters, cursor, context
>   File "/env/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1512, in _handle_dbapi_exception
>     sqlalchemy_exception, with_traceback=exc_info[2], from_=e
>   File "/env/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 178, in raise_
>     raise exception
>   File "/env/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1278, in _execute_context
>     cursor, statement, parameters, context
>   File "/env/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 593, in do_execute
>     cursor.execute(statement, parameters)
> sqlalchemy.exc.OperationalError: (psycopg2.errors.DeadlockDetected) deadlock detected
> DETAIL:  Process 16605 waits for AccessExclusiveLock on relation 16492 of database 16592; blocked by process 18557.
> Process 18557 waits for AccessShareLock on relation 16431 of database 16592; blocked by process 16605.
> HINT:  See server log for query details.
> 
> [SQL: 
> alter table agdc.dataset_source add column if not exists updated
> timestamptz not null default now();
> ]
> (Background on this error at: http://sqlalche.me/e/13/e3q8)

### Steps to reproduce the behaviour

Run `datacube system init` on an existing database with lots of records and some activity.

### Environment information

datacube 1.8.2.dev...

### PR DESCRIPTION
**Pull request by whatnick**
### Reason for this pull request
Adapt timestamping columns to run without locks on large databases

### Proposed changes
See details in [comments](https://github.com/opendatacube/datacube-core/issues/997#issuecomment-670347623).

 - [x] Closes #997
 - [x] Tests added / passed
 - [x] Fully documented, including `docs/about/whats_new.rst` for all changes

Please wait on tests on live DB's before merging this.

### CODE REVIEW

File: datacube/drivers/postgres/_core.py [hunk 1]
```diff
@@ -104,6 +107,8 @@ def ensure_db(engine, with_permissions=True):
             METADATA.create_all(c)
             _LOG.info("Creating triggers.")
             install_timestamp_trigger(c)
+            _LOG.info("Creating added column")
```
Comment (alexgleith) [inline datacube/drivers/postgres/_core.py]:
```suggestion
            _LOG.info("Creating added column.")
```

Comment (alexgleith) [reply datacube/drivers/postgres/_core.py]: Gots to be consistent ;-)

File: datacube/drivers/postgres/sql.py [hunk 2]
```diff
@@ -39,7 +39,12 @@ def visit_create_view(element, compiler, **kw):
 
 UPDATE_COLUMN_MIGRATE_SQL_TEMPLATE = """
 alter table {schema}.{table} add column if not exists updated
-timestamptz not null default now();
+timestamptz default null;
```
Comment (alexgleith) [inline datacube/drivers/postgres/sql.py:42]: Does this need to write rows? It's probably the default, defaulting to null anyway... Testing will show how it works :-)

**Review by alexgleith — CHANGES_REQUESTED** [review]
Looks good, @whatnick. Only a single extremely minor change!

File: datacube/drivers/postgres/_schema.py [hunk 3]
```diff
@@ -125,4 +125,7 @@
 
     # Note that the `updated` column is not included here to maintain backwards-compatibility
     # with pre-1.8.3 datacubes (and it is not used by any internal ODC functionality yet anyway)
+
+    # This table is append only and uses a migrations based `added` column to keep track of new
+    # dataset locations being added. The added column defaults to `now()`
```
Comment (jeremyh) [inline datacube/drivers/postgres/_schema.py]: The rows here can be removed too (via `index.datasets.remove_location()`), so they're not quite append-only. Just immutable.

File: datacube/drivers/postgres/sql.py (hunk 2 above)
Comment (jeremyh) [reply datacube/drivers/postgres/sql.py:42]: I was wondering the same thing. I assume it's the same as not mentioning a default... but we'll find out.

File: integration_tests/index/test_update_columns.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+def test_added_column(clirunner, uninitialised_postgres_db):
+    # Run on an empty database.
+    result = clirunner(["system", "init"])
+    assert "Created." in result.output
+
+    with uninitialised_postgres_db.connect() as connection:
+        assert check_column(connection, _schema.METADATA_TYPE.name, "updated")
+        assert check_column(connection, _schema.PRODUCT.name, "updated")
+        assert check_column(connection, _schema.DATASET.name, "updated")
+        assert check_column(connection, _schema.DATASET_LOCATION.name, "added")
```
Comment (jeremyh) [inline integration_tests/index/test_update_columns.py:94]: Now drop the column and rerun init :upside_down_face:

**Review by jeremyh — COMMENTED** [review]
I haven't run it, but it looks good to me

File: integration_tests/index/test_update_columns.py (hunk 4 above)
Comment (jeremyh) [reply integration_tests/index/test_update_columns.py:94]: I always feel slight paranoia when all tests are positive.

Maybe add an `assert not check_column(.... table, "fake_column")` so we know the function is working.

**Comment by whatnick** [comment]
Without changing tests this code has been successfully run on development `ows` database with live webservices consuming the datacube. Results are here:

image

image

image

**Review by alexgleith — APPROVED** [review]

**Review by jeremyh — APPROVED** [review]

**Pull request merged**